### DISCUSSION TITLE
Production Unable to locate a class or view for component [layouts.base]

### DISCUSSION THREAD

**Discussion by mattcookdev** [livewire/livewire#2165] [1. Help]
I develop on a M1 Macbook Pro 13", and everything works as intended. However, when deploying my Livewire application to a production server deployed by Laravel Forge, I get the error below.
```php
InvalidArgumentException

  Unable to locate a class or view for component [layouts.base].

  at vendor/laravel/framework/src/Illuminate/View/Compilers/ComponentTagCompiler.php:274
    270▕         if ($viewFactory->exists($view = $this->guessViewName($component))) {
    271▕             return $view;
    272▕         }
    273▕
  ➜ 274▕         throw new InvalidArgumentException(
    275▕             "Unable to locate a class or view for component [{$component}]."
    276▕         );
    277▕     }
    278▕
```

As stated, this error only happens on the forge server and not on my local development environment.

The component it is referring to is my layout component, and I do it pretty much identical to how Caleb does it in the screencasts. For reference;
```php
// app\View\Components\Layouts\App.php

<?php

namespace App\View\Components\Layouts;

use Illuminate\View\Component;

class App extends Component
{
    /**
     * Get the view / contents that represent the component.
     *
     * @return \Illuminate\View\View|string
     */
    public function render()
    {
        return view('layouts.app');
    }
}
```

```html
<!-- resources\views\layouts\app.blade.php -->

<x-layouts.base>
    <!-- contents -->
</x-layouts.base>
```

(Base is pretty much the same, just different blade template content.)

I was hoping someone might be able to provide some insight into why this happens and how I can go about fixing it. I've been trying to fix this for most of the day now and I'm completely out of ideas, and I've been unable to find anything of relevance on the internet other than it might have something to do with the operating system case sensitivity, but again, I'm not really sure how to fix that.

---

@calebporzio Sorry to tag you, but as stated above this is pretty much identical to the approach you take to layouts in the sponsor screencasts, so I was wondering if you have run into this before and have a solution, or if you perhaps have any idea why it's happening in general? Thanks! 😅

**Comment by ShekhSaifuddin007** [comment]
@zyxxyzyxz do like this..

```blade
<x-base-layout>
     .........
</x-base-layout>
```

**Reply by mattcookdev** [reply]
> @zyxxyzyxz do like this..
> 
> ```
> <x-base-layout>
>      .........
> </x-base-layout>
> ```

Hey @ShekhSaifuddin007 
Thanks for the suggestion but it didn't work. The issue as far as I'm aware is not with the code. It appears to be something happening on the server. Because this code works perfectly normal on my mac and only breaks when I deploy it on a Forge server.

So hoping someone else has run into this issue and knows a solution.
Also had some other very funky things happening with Livewire when deployed on a Forge server as well such as the html head tag getting really screwed up and inserting head contents like favicons into the body instead of inside the head tag when using a custom layout file... Same situation as above, works perfectly good on local Mac development env, just not the server. 
![image](https://user-images.githubusercontent.com/22277786/101193364-707af500-3697-11eb-8017-5345ec6605de.png)
(As you can see, all these meta tags and links should be in the head section and the div with flex-1 should be the only element inside the body, but instead the meta links are getting shoved in the body as well.) If I run *exactly* the same code on my mac, works totally fine. Even in production mode lol.

Very confusing and very frustrating problems, since it's basically impossible for me to reproduce them in my actual development environment (Which is what makes me think it's a server issue).

**Reply by stevecreekmore** [reply]
Problem solved! I missed the fact that Livewire references the the Views directory under App not the Views under Resources. They in affect redirect to the layouts under Resources/views/layouts.
![Solution](https://user-images.githubusercontent.com/5994720/102410625-070bb800-3fae-11eb-8321-8d04bac77b1f.png)

**Comment by joshhanley** [comment]
@zyxxyzyxz could potentially be a view caching issue. 

Make sure when you deploy to your server that you run on your production server `php artisan view:clear` which will clear all compiled blade views.

If the above doesn't work, make sure that app/View/Components/Layouts/Base.php is being deployed to your server properly, because it sounds like it's not finding that class.
And they try and run `composer dump-autoload`.

Hope this helps!

**Reply by mattcookdev** [reply]
Hey @joshhanley
Thanks for taking the time to reply. Sadly this isn't it. Through forge I run all those types of commands after deploy already, and this issue still happens. Also tried dump-autoload with no luck either. I've also tried running all the commands manually as well in-case forge was screwing up, but that didn't work either.

I'm trying to think about what other reasons it works on my mac but doesn't work on the server. Even if I completely wipe it from my mac, and set it up exactly the same way I would set it up on forge, it will work on the mac, but not on the server.

This has left me with theories that it's possibly nginx or Ubuntu related, but I'm really not sure at this point, because that doesn't really make any sense either. Someone said it could be to do with OS case sensitivity, but even if it is that, I'm not entirely sure how to fix that, and I've never had issues with Forge/Ubuntu/Livewire up until now, so I got a feeling something must have changed in one of them to be causing this.

**Reply by joshhanley** [reply]
@zyxxyzyxz hmm yeah I've run into something like that before, I'll look through my commit history to see if I can find anything.

In the meantime, can you post up exact code for the layouts base component class and blade file, along with the full path and filename of each, so I can see if I can notice any issues with casing/namespace/ or something else?

**Reply by stevecreekmore** [reply]
I'm having the same issue on my Mac and have tried all the suggestions above.
![Screen Shot 2020-12-16 at 1 23 51 PM](https://user-images.githubusercontent.com/5994720/102402426-f7867200-3fa1-11eb-89d3-6ad65e5f3cc7.png)

**Reply by joshhanley** [reply]
@Creekmore108 was it working before and now it has stopped? 

Where is your base.blade.php file saved, is it in `resources/views/layouts/base.blade.php`?

**Comment by leolll** [comment]
**Solved my similar issue**
I'm also using Livewire with Forge and Envoyer.  On my Forge server I got an error when running `php artisan view:cache`.

Error message:
```
In ComponentTagCompiler.php line 275:
                                                                  
Unable to locate a class or view for component [ui.icon.plus].
```

But on my Mac I wouldn't get any errors.  Removing `php artisan view:cache` from my deployment scripts allowed me to deploy but then I'd get a 500 error on the server when viewing the page that uses that component.

**Solution**
I had a `.gitignore_global` file that I must have created years ago from a template and it had the following rule:
```
# Icon must ends with two \r.
Icon
```
Anyway, I deleted that ignore rule, committed the icon blade files and now my issue is solved.

Hope it helps someone.

**Comment by HelgeSverre** [comment]
For future reference:

The folder name where you place your blade components needs to be lower case, it will work on mac and windows, but when you deploy to a linux machine, if you named your components directory "Components" it will not work properly.

**Reply by MohamedAladdin** [reply]
That solution is the one, thank you very much

**Reply by rahmani-djamel** [reply]
you save my life thank you

**Reply by nsubramkavin** [reply]
Thank you. Yes, it's a great catch, that solved my project's similar issue..
